src/StageStridedLoads.cpp

```diff
@@ -0,0 +1,366 @@
+                    found_loads[Key{op->name, base, stride, r->lanes, op->type, a, s}][offset].push_back(op);
+                }
+            }
+        }
+        IRVisitor::visit(op);
+    }
+
+    void visit(const For *op) override {
+        if (can_prove(op->extent > 0)) {
+            // The loop body definitely runs
+
+            // TODO: worry about different iterations of the loop somehow not
+            // providing the evidence we thought it did.
+            IRVisitor::visit(op);
+        } else {
+            const IRNode *child_scope = op->body.get();
+            parent_scope[child_scope] = scope;
+            ScopedValue<const IRNode *> bind(scope, child_scope);
+            IRVisitor::visit(op);
+        }
+    }
+
+    void visit(const IfThenElse *op) override {
+        op->condition.accept(this);
+        {
+            const IRNode *child_scope = op->then_case.get();
+            parent_scope[child_scope] = scope;
+            ScopedValue<const IRNode *> bind(scope, child_scope);
+            op->then_case.accept(this);
+        }
+        if (op->else_case.defined()) {
+            const IRNode *child_scope = op->else_case.get();
+            parent_scope[child_scope] = scope;
+            ScopedValue<const IRNode *> bind(scope, child_scope);
+            op->else_case.accept(this);
+        }
+    }
+
+    void visit(const Allocate *op) override {
+        // Provide a mapping from load nodes to paddable allocations they belong
+        // to.
+        ScopedBinding<const Allocate *> bind(allocation_scope, op->name, op);
+        IRVisitor::visit(op);
+    }
+
+    using IRVisitor::visit;
+};
+
+// Replace a bunch of load expressions in a stmt
+class ReplaceStridedLoads : public IRMutator {
+public:
+    std::map<const Load *, Expr> replacements;
+    std::map<const Allocate *, int> padding;
+
+protected:
+    Expr visit(const Load *op) override {
+        auto it = replacements.find(op);
+        if (it != replacements.end()) {
+            return mutate(it->second);
+        } else {
+            return IRMutator::visit(op);
+        }
+    }
+
+    Stmt visit(const Allocate *op) override {
+        auto it = padding.find(op);
+        Stmt s = IRMutator::visit(op);
+        if (it == padding.end()) {
+            return s;
+        } else {
+            op = s.as<Allocate>();
+            internal_assert(op);
+            return Allocate::make(op->name, op->type, op->memory_type,
+                                  op->extents, op->condition,
+                                  op->body, op->new_expr, op->free_function,
+                                  std::max(it->second, op->padding));
+        }
+    }
+
+    using IRMutator::visit;
+};
+
+}  // namespace
+
+Stmt stage_strided_loads(const Stmt &s) {
+    Stmt stmt = s;
+    // The following pass assumes that same variable name is the same variable,
+    // which is violated by the prior unrolling pass. Fix it up first.
+    stmt = uniquify_variable_names(stmt);
```

Comment:
Should we just run this immediately following `unroll_loops` in Lower.cpp? There have been bugs with violations of this before iirc (in the bounds stuff I did a few years ago).